### PR TITLE
Opt-In Support by ENV, File, Prompt, and CLI Arg

### PR DESCRIPTION
**Pull request by clintoncwolfe**
## Description

Implements opt-in/out via filesystem markers, environment variables, interactive prompts, and CLI arguments. The latter three indirectly interact with the Decision::File implementation to persist choices to disk in some use cases.

### Key points of review:

* The crux of the functionality is in `decision.rb#check_and_persist`, which contains the logic negotiating the priorities among the different ways of making a decision.

* The wording of the interactive prompt is currently as follows (this requires review from Product and likely Legal as well):
```
+---------------------------------------------+
            Chef Telemetry Opt-In
Optionally, you may choose to participate in
the Chef Telemetry program, which helps improve
Chef products by collecting data. View the
Privacy Policy at:
https://www.chef.io/privacy-policy/

Enable usage data collection (y/n)?
> Yes
Persisting telemetry opt-in decision...
✔ Enabled telemetry

CLI App goes here
```

## Related Issue

Implements #15

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: lib/chef/telemetry/decision.rb [hunk 1]
```diff
@@ -1,12 +1,91 @@
...
+          return @enabled = prompt_decision.prompt(dir, file_decision)
+        end
+
+        # Otherwise no decision has been made, default to runtime opt-out
+        logger.debug "Telemetry decision - no decision, defaulting to opt-out"
+        @enabled = false
+      end
+
+      def self.check_and_persist(opts = {})
+        new(opts).check_and_persist
```
Comment (marcparadise) [inline lib/chef/telemetry/decision.rb]: Should this accept a dir to pass into `#check_and_persist?` for local dir, and/or allow that as an option that we handle?

File: lib/chef/telemetry/decision/file.rb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+          candidates = []
+
+          # Include the user home directory ~/.chef
+          candidates << ChefConfig::PathHelper.home(".chef/#{DECISION_FILE}")
+
+          # TODO: include software installation directory, eg /opt/chef
+
+          # Include /etc/chef if on unix-like
+          candidates << "/etc/chef/#{DECISION_FILE}" unless ChefConfig.windows?
```
Comment (marcparadise) [inline lib/chef/telemetry/decision/file.rb]: Can we include C:/chef for windows, for consistency of "in the install directory" on both platforms?

File: spec/decision_spec.rb [hunk 3]
```diff
@@ -1,93 +1,201 @@
...
-      env["CD"] = env_pwd
-    else
-      env["PWD"] = env_pwd
-    end
+  def check_option_behavior(test_params = { should_be_enabled: false, should_persist: false, should_prompt: false } )
+    Dir.mktmpdir do |dir|
 
-    allow(ChefConfig::PathHelper).to receive(:home).with(".chef").and_return(File.join(home, ".chef"))
-  end
+      dec.file_decision.local_dir = dir
```
Comment (marcparadise) [inline spec/decision_spec.rb]: This should pass without setting `local_dir` explicitly, since check_and_persist will do do that - maybe better to test that after `check_and_persist` is done, `local_dir` is set to `dir`?

File: lib/chef/telemetry/decision.rb (hunk 1 above)
Comment (clintoncwolfe) [reply lib/chef/telemetry/decision.rb]: Thanks, that's a good catch - those args mismatch.

File: lib/chef/telemetry/decision/argument.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        end
+
+        def disable?
+          arg_seek("disable")
+        end
+
+        private
+
+        def arg_seek(what)
+          return true if argv.include?("--chef-telemetry=#{what}")
```
[the remark this answers is not in the document]
Comment (clintoncwolfe) [reply lib/chef/telemetry/decision/argument.rb]: My best answer is "I'm not sure."  Options are only "declared" if you're using a framework like Thor or mixlib-cli, etc. license-acceptance provides this facility (directly inspecting ARGV) *and also* provides a mixin that implements a Thor option.  inspec uses Thor, and so would certainly be interested in a Thor mixin.

I'm happy to add the mixins, if people think they are wanted.

File: lib/chef/telemetry/decision/file.rb (hunk 2 above)
Comment (clintoncwolfe) [reply lib/chef/telemetry/decision/file.rb]: Certainly, I just didn't know what path(s) to include... any others?

File: spec/decision_spec.rb (hunk 3 above)
Comment (clintoncwolfe) [reply spec/decision_spec.rb]: Yep, this is cruft - thanks!

**Comment by susanev** [comment]
@clintoncwolfe @kekaichinose id like to simplify the wording a bit, what would you think about..

`Share Data with Chef  

Chef would like to collect anonymized usage and diagnostic data to help improve your experience.
Privacy Policy: https://www.chef.io/privacy-policy/


Allow Chef to collect anonymized usage and diagnostic data (yes/no)?
> yes 

Saving data collection setting...
 <check> Data collection setting saved.`

..

also, are they able to change this later? if so i want to edit the into paragraph text to make that clear.

**Comment by kekaichinose** [comment]
> @clintoncwolfe @kekaichinose id like to simplify the wording a bit, what would you think about..
> 
> ```
> Share Data with Chef  
> 
> Chef would like to collect anonymized usage and diagnostic data to 
> help improve your experience.
> Privacy Policy: https://www.chef.io/privacy-policy/
> 
> Allow Chef to collect anonymized usage and diagnostic data (yes/no)?
> > yes 
> 
> Saving data sharing setting...
>  <check> Data sharing setting saved.
> ```
> 
> ..also, are they able to change this setting later? if so i want to edit the intro paragraph text to make that clear.

Yes, there are a couple ways they can do that, which we could make clear in documentation. Then we could put the link to the documentation in the intro paragraph. Would that work, @susanev ?

To be clear, I think having another prompt to set disabling of telemetry probably is overkill and doesn't make sense here, right...?

**Comment by susanev** [comment]
@kekaichinose we need to at least inform them that they can change it later in this flow, but if you want to only explain how to do that in the documentation i guess i support that.

**Comment by clintoncwolfe** [comment]
Yes, they can change the setting later in a few ways...

Temporarily (this invocation only):
 * Run the program with the option `--chef-telemetry=disable`.
 * Run the program with the environment variable `CHEF_TELEMETRY=OPT_OUT_NO_PERSIST`

Persistently:
 * Run the app again, with `CHEF_TELEMETRY=OPT_OUT`. That will re-write the telemetry file to persist an opt-out choice.
 * Delete the file `{/etc/chef|C:\chef\}telemetry_options.yaml` and re-run the program. The user will be prompted again and may answer "no".
 * Edit the above file and change the `:enabled: true` value to false

Likewise, a use may be opted-out, but may opt-in for single invocation using `--chef-telemetry=enable` or `CHEF_TELEMETRY=OPT_IN_NO_PERSIST` for, perhaps, customer support purposes.

File: lib/chef/telemetry/decision/file.rb (hunk 2 above)
Comment (clintoncwolfe) [reply lib/chef/telemetry/decision/file.rb]: I've reworked this to default to using the home directory of the user, like license-acceptance. I've left a comment for future work in which we might add in per-product directories, but that might actually turn it into a globbing operation - there are a lot of variants due to the omnibus distros.

File: lib/chef/telemetry/decision.rb [hunk 5]
```diff
@@ -1,12 +1,91 @@
...
+        # If a non-persisting decision is made by env, only set runtime decision
+        logger.debug "Telemetry decision examining ephemeral ENV checks"
+        return @enabled = true if @env_decision.opt_in_no_persist?
+        return @enabled = false if @env_decision.opt_out_no_persist?
+
+        # Check to see if a persistent decision is made by env but not yet persisted
+        #  then persist it and set runtime decision
+        logger.debug "Telemetry decision examining persistent ENV checks"
+        if env_decision.opt_in?
+          file_decision.persist(true, dir) if !persisted? || file_decision.opt_out?
```
Comment (tyler-ball) [inline lib/chef/telemetry/decision.rb]: I want to make sure I am understanding this logic. If we set opt-in and persist with an environment variable, we persist that value if 1) it has not been persisted before or 2) if they previously had an opt-out decision recorded in the file.

So I'm understanding that the environment variable opt-in will overwrite the previous file-based opt-out, and it will persist that new value, correct?

File: lib/chef/telemetry/decision/argument.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (tyler-ball) [reply lib/chef/telemetry/decision/argument.rb]: Without using the mixin, the help would be missing `--chef-telemetry` correct?

File: lib/chef/telemetry/decision/argument.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
+class Chef
+  class Telemetry
+    class Decision
```
Comment (tyler-ball) [inline lib/chef/telemetry/decision/argument.rb]: I like the `Decision` class name, I think it is a little more clear than my usage of `Strategy`. `Strategy` is an overloaded, pattern-ey thing in my head.

File: lib/chef/telemetry/decision/prompt.rb [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+          @output = cfg[:output]
+          @input = cfg[:input] || STDIN
+        end
+
+        def prompt(dir, persistor)
+          logger.debug "Prompting for opt-in/out..."
+
+          output.puts <<~EOM
+            #{BORDER}
+            Share Data with Chef
```
Comment (tyler-ball) [inline lib/chef/telemetry/decision/prompt.rb]: Maybe a `:` at the end of this line since it is a header for the following sentence?

File: lib/chef/telemetry/decision/prompt.rb [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+            output.puts "\nPrompt timed out. Opting out of usage data\nfor this run."
+            # Do not opt-in on timeout
+            return false
+          }
+
+          if ChefConfig.windows?
+            # On windows, Timeout hangs on STDIN. TTY::Prompt's keypress is safe, but we
+            # can't distinguish a "timeout default" from a "they just pressed enter default"
+            # So, assume all defaults are opt-out.
+            answer = prompt.keypress("y/n?", default: "t", timeout: timeout)
```
Comment (tyler-ball) [inline lib/chef/telemetry/decision/prompt.rb]: Earlier it was prompting for `yes/no` but here it is `y/n` - lets bring those in line (unless there is a reason I am not understanding?)

**Comment by susanev** [comment]
@kekaichinose @clintoncwolfe, i collaborated a bit with @jonong1972 to make the bottom part less confusing.. could we have dynamic text there?

```
Turning on data sharing...
 <check> Data sharing turned on.
```

```
 Turning off data sharing...
 <check> Data sharing turned off.
```

**Comment by jonong1972** [comment]
> @kekaichinose @clintoncwolfe, i collaborated a bit with @jonong1972 to make the bottom part less confusing.. could we have dynamic text there?
> 
> ```
> Turning on data sharing...
>  <check> Data sharing turned on.
> ```
> 
> ```
>  Turning off data sharing...
>  <check> Data sharing turned off.
> ```

Yeah, this makes more sense. It is more in context with the action the user is taking.  When a user says "No" to Enable Data Usage - "Data sharing setting saved." is not what is expected. Makes more that whatever they agree or disagree agree to, should be the messaging.

File: lib/chef/telemetry/decision.rb (hunk 5 above)
Comment (clintoncwolfe) [reply lib/chef/telemetry/decision.rb]: That is correct, for certain values of environment variables. The -no-persist variants do not persist.

All code within `check_and_persist` is totally open to revision per the product team, anyway - I expect changes in logic here so ... just looking for someone to say what's what.

File: lib/chef/telemetry/decision/argument.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (clintoncwolfe) [reply lib/chef/telemetry/decision/argument.rb]: Yes, that's right. I've added a mixin for thor and mixlib-cli and will push that later today.

**Comment by tas50** [comment]
@clintoncwolfe I am very interested in the CLI option / ENV var portion of this work, but I'm not sold on the need for interactive prompting later. Our license agreement allows us to collect data and we should be prompting that tools will have telemetry enabled within that license acceptance logic. The assumption going forward is that everything would include telemetry and everything would use it if enabled. License acceptance occurring would be the trigger for that unless disabled explicitly by the user. That would be disabling it via CLI flag, ENV var, or yet to be build GUI configs.
